addons/binding/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/handler/SamsungTvHandler.java

```diff
@@ -0,0 +1,343 @@
+import org.jupnp.UpnpService;
+import org.jupnp.model.meta.LocalDevice;
+import org.jupnp.model.meta.RemoteDevice;
+import org.jupnp.registry.Registry;
+import org.jupnp.registry.RegistryListener;
+import org.openhab.binding.samsungtv.config.SamsungTvConfiguration;
+import org.openhab.binding.samsungtv.internal.service.ServiceFactory;
+import org.openhab.binding.samsungtv.internal.service.api.SamsungTvService;
+import org.openhab.binding.samsungtv.internal.service.api.ValueReceiver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link SamsungTvHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ * 
+ * @author Pauli Anttila - Initial contribution
+ */
+public class SamsungTvHandler extends BaseThingHandler implements
+		DiscoveryListener, RegistryListener, ValueReceiver {
+
+	private Logger logger = LoggerFactory.getLogger(SamsungTvHandler.class);
+
+	/** Polling job for searching UPnP devices on startup */
+	private ScheduledFuture<?> pollingJob;
+
+	/** Global configuration for Samsung TV Thing */
+	private SamsungTvConfiguration configuration;
+
+	private UpnpIOService upnpIOService;
+	private DiscoveryServiceRegistry discoveryServiceRegistry;
+	private UpnpService upnpService;
+
+	/** Samsung TV services */
+	private List<SamsungTvService> services;
+
+	private boolean powerOn = false;
+
+	public SamsungTvHandler(Thing thing, UpnpIOService upnpIOService,
+			DiscoveryServiceRegistry discoveryServiceRegistry,
+			UpnpService upnpService) {
+
+		super(thing);
+
+		logger.debug("Create a Samsung TV Handler for thing '{}'", getThing()
+				.getUID());
+
+		if (upnpIOService != null) {
+			this.upnpIOService = upnpIOService;
+		} else {
+			logger.debug("upnpIOService not set.");
+		}
+
+		if (discoveryServiceRegistry != null) {
+			this.discoveryServiceRegistry = discoveryServiceRegistry;
+			this.discoveryServiceRegistry.addDiscoveryListener(this);
+		}
+
+		if (upnpService != null) {
+			this.upnpService = upnpService;
+			this.upnpService.getRegistry().addListener(this);
+		} else {
+			logger.debug("upnpService not set.");
+		}
+
+		services = new ArrayList<>();
+	}
+
+	@Override
+	public void handleCommand(ChannelUID channelUID, Command command) {
+		logger.debug("Received channel: {}, command: {}", channelUID, command);
+
+		if (getThing().getStatus() == ThingStatus.ONLINE) {
+
+			// Delegate command to correct service
+
+			String channel = channelUID.getId();
+
+			for (SamsungTvService service : services) {
+				if (service != null) {
+					List<String> supportedCommands = service
+							.getSupportedChannelNames();
+					for (String s : supportedCommands) {
+						if (channel.equals(s)) {
+							service.handleCommand(channel, command);
+							return;
+						}
+					}
+				}
+			}
+
+			logger.warn("Channel '{}' not supported", channelUID);
+		} else {
+			logger.warn("Samsung TV '{}' is OFFLINE", getThing().getUID());
```

Comment:
I would reduce this to debug level as it can otherwise clutter the log files. We will make sure that there is an easy way to be informed if things are unexpectedly offline.